src/test/test-alloc-util.c

```diff
@@ -23,18 +24,26 @@ static void test_GREEDY_REALLOC(void) {
         _cleanup_free_ int *a = NULL, *b = NULL;
         size_t n_allocated = 0, i;
 
-        /* Give valgrind a chance to verify our realloc operations */
+        /* Give valgrind a chance to verify our realloc() operations */
 
-        for (i = 0; i < 2048; i++) {
+        for (i = 0; i < 20480; i++) {
                 assert_se(GREEDY_REALLOC(a, n_allocated, i + 1));
+                assert_se(n_allocated >= i + 1);
+                assert_se(malloc_usable_size(a) >= (i + 1) * sizeof(int));
                 a[i] = i;
```

Comment:
After this loop performs the reallocs, a separate loop should verify that all of the array elements are the expected value. This would (hopefully) catch any issues like the one before where some memory doesn't get properly copied over.